testutils/test_server.go

```diff
@@ -0,0 +1,228 @@
+func (ts *TestServer) Server() *tchannel.Channel {
+	return ts.server
+}
+
+// ServerState is a convenience method to introspect the runtime state of the
+// server channel.
+func (ts *TestServer) ServerState(opts *tchannel.IntrospectionOptions) *tchannel.RuntimeState {
+	return ts.server.IntrospectState(opts)
+}
+
+// HostPort returns the host:port for clients to connect to. Note that this may
+// not be the same as the host:port of the server channel.
+func (ts *TestServer) HostPort() string {
+	// Use this wrapper to enable registering handlers on the server, but
+	// connecting to a relay.
+	return ts.server.PeerInfo().HostPort
+}
+
+// ServiceName returns the service name of the server channel.
+func (ts *TestServer) ServiceName() string {
+	return ts.server.PeerInfo().ServiceName
+}
+
+// Register registers a handler on the server channel.
+func (ts *TestServer) Register(h tchannel.Handler, methodName string) {
+	ts.server.Register(h, methodName)
+}
+
+// CloseAndVerify closes all channels, then verifies that no message exchanges
+// or goroutines were leaked.
+func (ts *TestServer) CloseAndVerify() {
+	ts.close()
+	ts.verify()
+}
+
+// NewClient is a convenience wrapper around testutils.NewClient.
+func (ts *TestServer) NewClient(opts *ChannelOpts) *tchannel.Channel {
+	return NewClient(ts, opts)
+}
+
+func (ts *TestServer) close() {
+	ts.server.Close()
+}
+
+func (ts *TestServer) verify() {
+	ts.waitForChannelClose(ts.server)
+
+	ts.verifyExchangesCleared()
+	ts.verifyNoGoroutinesLeaked()
+}
+
+func (ts *TestServer) waitForChannelClose(ch *tchannel.Channel) {
+	if ts.Failed() {
+		return
+	}
+	started := time.Now()
+
+	var state tchannel.ChannelState
+	for i := 0; i < 50; i++ {
+		if state = ch.State(); state == tchannel.ChannelClosed {
+			return
+		}
+
+		runtime.Gosched()
+		if i < 5 {
+			continue
+		}
+
+		sleepFor := time.Duration(i) * 100 * time.Microsecond
+		time.Sleep(Timeout(sleepFor))
+	}
+
+	// Channel is not closing, fail the test.
+	sinceStart := time.Since(started)
+	ts.Errorf("Channel did not close after %v, last state: %v", sinceStart, state)
+}
+
+func (ts *TestServer) verifyExchangesCleared() {
+	if ts.Failed() {
+		return
+	}
+	// Ensure that all the message exchanges are empty.
+	serverState := ts.server.IntrospectState(&tchannel.IntrospectionOptions{
+		IncludeExchanges: true,
+	})
+	if exchangesLeft := describeLeakedExchanges(serverState); exchangesLeft != "" {
+		ts.Errorf("Found uncleared message exchanges on server:\n%v", exchangesLeft)
+	}
+}
+
+func (ts *TestServer) verifyNoGoroutinesLeaked() {
+	leaks, err := goroutines.IdentifyLeaks(ts.verifyOpts)
+	if err != nil {
+		ts.Error(err.Error())
+		return
+	}
+	// Count the leaked goroutines, even if this test has already failed.
+	if prevLeaks := _previouslyLeakedGoroutines.Add(uint32(len(leaks))); prevLeaks > 0 {
+		ts.Log("Skipping check for leaked goroutines, since we've already leaked at least %v.", prevLeaks)
```

Comment:
log says skipping check, but there's no return here. also message seems a little abrupt, "since we've already leaked at least 5". maybe "Skipping check for leaked goroutines due to %v previous leaks"?